### PR TITLE
AB - New Session Pool

### PR DESCRIPTION
**Pull request by anup-honey**
Fixes #<issue_number_goes_here> (it's a good idea to open an issue first for discussion)

- [x] Tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

### CODE REVIEW

File: src/session-pool.js [hunk 1]
```diff
@@ -109,221 +106,77 @@ function SessionPool(database, options) {
...
+    {
+      create: this.createReadSession.bind(this),
+      destroy: this.destroySession.bind(this),
+    },
+    {
+      max: this.options.maxReads,
+      min: this.options.minReads,
+      acquireTimeoutMillis,
+      autostart: false,
+      maxWaitingClients: this.options.fail ? 0 : 100,
```
Comment (walshie4) [inline src/session-pool.js]: we should make this 100 configurable

File: src/session-pool.js [hunk 2]
```diff
@@ -109,221 +106,77 @@ function SessionPool(database, options) {
...
+      destroy: this.destroySession.bind(this),
+    },
+    {
+      max: this.options.maxWrites,
+      min: this.options.minWrites,
+      acquireTimeoutMillis,
+      autostart: false,
+      maxWaitingClients: this.options.fail ? 0 : 100,
+      idleTimeoutMillis: this.options.idlesAfter * 60000,
+    }
```
Comment (walshie4) [inline src/session-pool.js]: I think we should allow access to the other `node-pool` configuration options.  Some of them seem useful.

https://github.com/coopernurse/node-pool#createpool

File: src/session-pool.js [hunk 3]
```diff
@@ -377,153 +230,53 @@ SessionPool.prototype.requestStream = function(config) {
...
  */
-SessionPool.prototype.size = function() {
-  return this.available() + this.borrowed();
+SessionPool.prototype.getReadSession = function() {
+  const self = this;
+
+  if (!self.isOpen) {
+    return Promise.reject(new Error('Database is closed.'));
+  }
+  return self.requestQueue_.add(() => self.readPool.acquire());
```
Comment (walshie4) [inline src/session-pool.js]: Can we remove these anonymous wrapper functions?

```javascript
return self.requestQueue_.add(self.readPool.acquire);
```

File: src/session-pool.js [hunk 4]
```diff
@@ -109,221 +106,77 @@ function SessionPool(database, options) {
...
+      min: this.options.minWrites,
+      acquireTimeoutMillis,
+      autostart: false,
+      maxWaitingClients: this.options.fail ? 0 : 100,
+      idleTimeoutMillis: this.options.idlesAfter * 60000,
+    }
+  );
 
-  this.traces_ = new Map();
+  this.getStats();
```
Comment (walshie4) [inline src/session-pool.js]: I don't think this call makes sense now that this function doesn't set-up logging stats.

File: src/session-pool.js (hunk 3 above)
Comment (walshie4) [reply src/session-pool.js]: Nvm, this will break things

File: src/session-pool.js [hunk 5]
```diff
@@ -16,56 +16,53 @@
...
   idlesAfter: 10,
   keepAlive: 50,
-  max: 100,
-  maxIdle: 1,
-  min: 0,
-  writes: 0,
+  maxReads: 100,
+  maxWrites: 100,
+  minReads: 10,
+  minWrites: 5,
```
Comment (DuaneGarber) [inline src/session-pool.js]: Are these minRead/Write defaults too opinionated? Should they be 0?

File: package.json [hunk 6]
```diff
@@ -68,13 +68,14 @@
     "@google-cloud/common-grpc": "^0.6.0",
     "array-uniq": "^1.0.3",
     "arrify": "^1.0.1",
+    "bluebird": "^3.5.1",
```
Comment (walshie4) [inline package.json]: Do we actually leverage bluebird features?

Comment (anup-honey) [reply package.json]: Promise.try

Comment (walshie4) [reply package.json]: I don't see this anywhere, do you have a link?

Comment (walshie4) [reply package.json]: That may also be a small enough use-case that we may want to avoid bringing this in

File: src/session-pool.js [hunk 7]
```diff
@@ -109,221 +98,69 @@ function SessionPool(database, options) {
...
-/**
- * Checks to see if the pool is full by calculating the current size and the
- * pending number of sessions being created.
- *
- * @return {boolean}
- */
-SessionPool.prototype.isFull = function() {
-  return this.size() + this.pendingCreates_ >= this.options.max;
-};
+  EventEmitter.call(this);
```
Comment (walshie4) [inline src/session-pool.js]: this needs to go away if we're removing EventEmitter

**Pull request merged**